language/diem-framework/modules/NetworkIdentity.move

```diff
@@ -140,25 +139,167 @@ module DiemFramework::NetworkIdentity {
+    /// the uniqueness of its elements are guaranteed, before and after the bulk
+    /// removal. The `to_remove` argument, on the other hand, does not guarantee
+    /// to be a set and hence can have duplicated elements.
+    fun remove_members_internal<T: drop>(
+        members: &mut vector<T>,
+        to_remove: &vector<T>,
+    ): bool {
+        let num_existing = Vector::length(members);
+        let num_to_remove = Vector::length(to_remove);
 
-    spec remove_identity {
+        let i = 0;
+        while ({
+            spec {
+                assert i <= num_to_remove;
+                // the set can never grow in size
+                assert len(members) <= len(old(members));
+                // the current set maintains the uniqueness of the elements
+                assert forall j in 0..len(members), k in 0..len(members):
+                    members[j] == members[k] ==> j == k;
+                // all elements in the the current set come from the original set
+                assert forall j in 0..len(members): contains(old(members), members[j]);
+                // the current set never contains anything from the `to_remove` vector
+                assert forall j in 0..i: !contains(members, to_remove[j]);
+                // the current set should never remove an element from the original set which is not in `to_remove`
+                assert forall j in 0..len(old(members)): (contains(to_remove[0..i], old(members)[j]) || contains(members, old(members)[j]));
+                // having the same member means that all elements in the `to_remove` vector we have seen so far are not
+                // in the existing set, and vice versa.
+                assert len(members) == len(old(members)) <==> (forall j in 0..i: !contains(old(members), to_remove[j]));
+            };
+            (i < num_to_remove)
+        }) {
+            let entry = Vector::borrow(to_remove, i);
+            let (exist, index) = Vector::index_of(members, entry);
+            if (exist) {
+                Vector::swap_remove(members, index);
+            };
+            i = i + 1;
+        };
+
+        Vector::length(members) < num_existing
+    }
+    spec remove_members_internal {
+        pragma opaque;
+        // TODO(mengxu): this is to force the prover to honor the "opaque" pragma in the ignore opaque setting
+        aborts_if [abstract] false;
+        aborts_if [concrete] false;
+        include RemoveMembersInternalEnsures<T> {
+            old_members: old(members),
+            new_members: members,
+        };
+        // ensures that the `members` argument is and remains a set
+        invariant forall i in 0..len(members), j in 0..len(members):
+            members[i] == members[j] ==> i == j;
+        // returns whether an element is removed from the set
+        ensures result == (exists i in 0..len(to_remove): contains(old(members), to_remove[i]));
+    }
+    spec schema RemoveMembersInternalEnsures<T: drop> {
```

Comment:
Can you not use these in the loop invariants?

Reply:
Actually we can, I guess I did not use it just because of convention...

Reply:
Okay interesting, just figure why write it twice